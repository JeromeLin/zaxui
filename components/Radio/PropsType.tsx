interface BaseRadioProps {
  theme?: 'default' | 'primary' | 'info' | 'success' | 'warning' | 'error';
  shape?: 'radius' | 'round';
  size?: 'lg';
  type?: 'button' | 'cell';
  disabled?: boolean;
  block?: boolean;
  onChange?: (checked: boolean) => void;
  checked?: boolean;
  defaultChecked?: boolean;
  value?: number | string;
}

export interface RadioProps extends BaseRadioProps {
  prefixCls?: string;
  className?: string;
}

interface BaseRadioGroupProps {
  theme?: 'default' | 'primary' | 'info' | 'success' | 'warning' | 'error';
  shape?: 'radius' | 'round';
  size?: 'lg';
  type?: 'button' | 'cell';
  disabled?: boolean;
  block?: boolean;
  onChange?: (checked: boolean) => void;
  value?: number | string;
  defaultValue?: number | string;
  compact?: boolean;
}

export interface RadioGroupProps extends BaseRadioGroupProps {
  prefixCls?: string;
  className?: string;
}