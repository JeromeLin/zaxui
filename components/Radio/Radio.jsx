import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Cell from '../Cell';
import Button from '../Button';
import Icon from '../Icon';

function getChecked(props, defaultChecked) {
  if ('checked' in props && props.checked) {
    return props.checked;
  }
  if ('defaultChecked' in props && props.defaultChecked) {
    return props.defaultChecked;
  }
  return defaultChecked;
}

class Radio extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      checked: getChecked(props, false),
    };
    this.onValueChange = this.onValueChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: !!nextProps.checked,
      });
    }
  }

  onValueChange() {
    const { disabled, onChange } = this.props;

    if (disabled) return;

    const checked = true;
    this.setState({ checked });
    typeof onChange === 'function' && onChange(checked);
  }

  render() {
    const { prefixCls, className, type, theme, block, value, disabled, children } = this.props;
    const { checked } = this.state;

    const cls = classnames({
      [`${prefixCls}`]: true,
      [className]: !!className,
    });

    if (type === 'cell') {
      return (
        <Cell disabled={disabled} description={checked ? <Icon type="right" theme={disabled ? null : theme} /> : null} onClick={() => {}}>
          <input type="radio" className={`${prefixCls}-input`} value={value} disabled={disabled} checked={checked} onChange={this.onValueChange} />
          {children}
        </Cell>
      );
    }

    return (
      <Button className={cls} theme={theme} size="xs" block={block} bordered={!checked} disabled={disabled}>
        <input type="radio" className={`${prefixCls}-input`} value={value} disabled={disabled} checked={checked} onChange={this.onValueChange} />
        {children}
      </Button>
    );
  }
}

Radio.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  type: PropTypes.oneOf(['button', 'cell']),
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool, // eslint-disable-line
  disabled: PropTypes.bool,
  block: PropTypes.bool,
  onChange: PropTypes.func,
};

Radio.defaultProps = {
  prefixCls: 'za-radio',
  disabled: false,
  block: false,
};

export default Radio;
