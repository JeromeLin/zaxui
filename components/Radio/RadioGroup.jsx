import React, { PureComponent, cloneElement } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

function getCheckedValue(children) {
  let checkedValue = null;
  React.Children.forEach(children, (radio) => {
    if (radio.props && radio.props.checked) {
      checkedValue = radio.props.value;
    }
  });
  return checkedValue;
}

class RadioGroup extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      value: props.value || props.defaultValue || getCheckedValue(props.children),
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps || getCheckedValue(nextProps.children)) {
      this.setState({
        value: nextProps.value || nextProps.defaultValue || getCheckedValue(nextProps.children),
      });
    }
  }

  onRadioChange(value) {
    this.setState({ value });
    this.props.onChange(value);
  }

  render() {
    const { prefixCls, className, theme, shape, type, block, disabled, gather, children } = this.props;

    const items = React.Children.map(children, (element, index) => {
      return cloneElement(element, {
        key: index,
        type,
        theme,
        block: block || element.props.block,
        disabled: disabled || element.props.disabled,
        onChange: () => this.onRadioChange(element.props.value),
        // use '==' because the result will fail when the value's typeof is Number
        checked: (this.state.value == element.props.value), // eslint-disable-line
      });
    });

    const cls = classnames({
      [`${prefixCls}`]: true,
      [className]: !!className,
      [`shape-${shape}`]: !!shape,
      'is-gather': gather,
      block,
      disabled,
    });

    return (
      <div className={cls}>
        {items}
      </div>
    );
  }
}

RadioGroup.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  theme: PropTypes.oneOf(['default', 'primary', 'info', 'success', 'warning', 'error']),
  type: PropTypes.oneOf(['button', 'cell']),
  shape: PropTypes.oneOf(['radius', 'round']),
  block: PropTypes.bool,
  disabled: PropTypes.bool,
  gather: PropTypes.bool,
  onChange: PropTypes.func,
};

RadioGroup.defaultProps = {
  prefixCls: 'za-radio-group',
  className: null,
  theme: 'primary',
  type: null,
  shape: null,
  block: false,
  disabled: false,
  gather: false,
  onChange() {},
};

export default RadioGroup;
