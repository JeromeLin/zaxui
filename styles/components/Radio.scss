$prefixCls: za-radio;

.#{$prefixCls} {
  position: relative;
  margin-left: r(10);

  &:first-child {
    margin-left: 0;
  }

  &-input {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
  }

  &-group {
    &.block {
      display: flex;

      /* 通栏模式字体增大 */
      .#{$prefixCls} {
        font-size: r(15);
      }
    }

    &.shape-radius {
      .#{$prefixCls} {
        border-radius: $radius-md;
      }
    }

    &.shape-round {
      .#{$prefixCls} {
        border-radius: r(15);
      }
    }

    &.is-compact {
      .#{$prefixCls} {
        margin-left: -1px;
      }

      &.shape-radius {
        .#{$prefixCls} {
          border-radius: 0;

          &:first-child {
            border-radius: $radius-md 0 0 $radius-md;
          }

          &:last-child {
            border-radius: 0 $radius-md $radius-md 0;
          }
        }
      }

      &.shape-round {
        .#{$prefixCls} {
          border-radius: 0;

          &:first-child {
            border-radius: r(15) 0 0 r(15);
          }

          &:last-child {
            border-radius: 0 r(15) r(15) 0;
          }
        }
      }
    }
  }
}
