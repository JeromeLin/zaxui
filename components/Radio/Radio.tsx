import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { RadioProps } from './PropsType';
import Cell from '../Cell';
import Button from '../Button';
import Icon from '../Icon';

function getChecked(props, defaultChecked) {
  if ('checked' in props && props.checked) {
    return props.checked;
  }
  if ('defaultChecked' in props && props.defaultChecked) {
    return props.defaultChecked;
  }
  return defaultChecked;
}

export { RadioProps };

export default class Radio extends PureComponent<RadioProps, any> {

  static Group: any;

  static defaultProps = {
    prefixCls: 'za-radio',
    theme: 'primary',
    disabled: false,
    block: false,
  }

  constructor(props) {
    super(props);
    this.state = {
      checked: getChecked(props, false),
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('checked' in nextProps) {
      this.setState({
        checked: !!nextProps.checked,
      });
    }
  }

  onValueChange = () => {
    const { disabled, onChange } = this.props;

    if (disabled) return;

    const checked = true;
    this.setState({ checked });
    typeof onChange === 'function' && onChange(checked);
  }

  render() {
    const { prefixCls, className, type, theme, shape, block, value, disabled, children } = this.props;
    const { checked } = this.state;

    const cls = classnames(`${prefixCls}`, className, {
      [`theme-${theme}`]: !!theme,
      [`shape-${shape}`]: !!shape,
      checked,
      disabled,
    });

    if (type === 'cell') {
      return (
        <Cell disabled={disabled} description={checked ? <Icon type="right" theme={disabled ? undefined : theme} /> : null} onClick={() => {}}>
          <input type="radio" className={`${prefixCls}-input`} value={value} disabled={disabled} checked={checked} onChange={this.onValueChange} />
          {children}
        </Cell>
      );
    }

    if (type === 'button') {
      return (
        <Button className={cls} theme={theme} shape={shape} size="xs" block={block} bordered={!checked} disabled={disabled}>
          <input type="radio" className={`${prefixCls}-input`} value={value} disabled={disabled} checked={checked} onChange={this.onValueChange} />
          {children}
        </Button>
      );
    }

    return (
      <div className={cls}>
        <div className={`${prefixCls}-wrapper`}>
          <span className={`${prefixCls}-inner`} />
          {children && <span className={`${prefixCls}-text`}>{children}</span>}
          <input type="radio" className={`${prefixCls}-input`} disabled={disabled} checked={checked} onChange={this.onValueChange} />
        </div>
      </div>
    );
  }
}

