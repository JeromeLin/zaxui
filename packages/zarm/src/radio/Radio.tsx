import * as React from 'react';
import classnames from 'classnames';
import type { BaseRadioProps } from './interface';
import RadioGroup from './RadioGroup';
import Cell from '../cell';

const getChecked = (props: RadioProps, defaultChecked: boolean) => {
  if (typeof props.checked !== 'undefined') {
    return props.checked;
  }
  if (typeof props.defaultChecked !== 'undefined') {
    return props.defaultChecked;
  }
  return defaultChecked;
};

type RadioSpanProps = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;
type RadioCellProps = Omit<
  React.HTMLAttributes<HTMLDivElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;
type RadioButtonProps = Omit<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  'type' | 'defaultChecked' | 'checked' | 'value' | 'onChange'
>;

export type RadioProps = Partial<RadioSpanProps & RadioCellProps & RadioButtonProps> &
  BaseRadioProps & {
    prefixCls?: string;
    onChange?: (e?: React.ChangeEvent<HTMLInputElement>) => void;
  };

interface CompoundedComponent
  extends React.ForwardRefExoticComponent<RadioProps & React.RefAttributes<HTMLElement>> {
  Group: typeof RadioGroup;
}

const Radio = React.forwardRef<unknown, RadioProps>(
  (
    {
      prefixCls = 'za-radio',
      className,
      type,
      value,
      checked,
      shape,
      defaultChecked,
      disabled = false,
      id,
      children,
      onChange,
      ...rest
    }: RadioProps,
    ref,
  ) => {
    const radioRef = (ref as any) || React.createRef<HTMLElement>();
    const [currentCheck, setCurrentCheck] = React.useState<boolean>(
      getChecked({ checked, defaultChecked }, false),
    );

    const onValueChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      if (disabled) {
        return;
      }
      const newChecked = !currentCheck;
      if (typeof checked === 'undefined') {
        setCurrentCheck(newChecked);
      }

      if (typeof onChange === 'function') {
        onChange(e);
      }
    };

    const cls = classnames(prefixCls, className, {
      [`${prefixCls}--checked`]: currentCheck,
      [`${prefixCls}--disabled`]: disabled,
      [`${prefixCls}--untext`]: !children,
    });

    const inputRender = (
      <input
        id={id}
        type="radio"
        className={`${prefixCls}__input`}
        value={value}
        disabled={disabled}
        checked={currentCheck}
        onChange={onValueChange}
      />
    );

    React.useEffect(() => {
      setCurrentCheck(getChecked({ checked, defaultChecked }, false));
    }, [checked, defaultChecked]);

    const radioRender = (
      <span className={cls} ref={radioRef} {...(rest as RadioSpanProps)}>
        <span className={`${prefixCls}__widget`}>
          <span className={`${prefixCls}__inner`} />
        </span>
        {children && <span className={`${prefixCls}__text`}>{children}</span>}
        {inputRender}
      </span>
    );

    if (type === 'cell') {
      return (
        <Cell
          disabled={disabled}
          className={className}
          onClick={() => {}}
          {...(rest as RadioCellProps)}
        >
          {radioRender}
        </Cell>
      );
    }

    if (type === 'button') {
      return (
        <button
          type="button"
          disabled={disabled}
          className={cls}
          ref={radioRef}
          {...(rest as RadioButtonProps)}
        >
          {children}
          {inputRender}
        </button>
      );
    }

    return radioRender;
  },
) as CompoundedComponent;

Radio.displayName = 'Radio';
Radio.defaultProps = {
  prefixCls: 'za-radio',
  disabled: false,
  shape: 'radius',
};

export default Radio;
