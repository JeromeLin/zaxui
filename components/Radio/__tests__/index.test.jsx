import React from 'react';
import { render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Radio from '../index';

describe('Radio', () => {
  // 普通
  it('renders correctly', () => {
    const wrapper = render(
      <Radio.Group value="0" onChange={jest.fn()}>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // 指定默认值
  it('defaultValue', () => {
    const wrapper = render(
      <Radio.Group defaultValue="1">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.find('input[type="radio"][value="1"]').prop('checked')).toEqual(true);
  });

  // 指定默认值
  it('disabled', () => {
    const wrapper = render(
      <Radio.Group>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2" disabled>选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.find('input[type="radio"][value="2"]').prop('disabled')).toEqual(true);
  });

  // 圆角
  it('shape is radius', () => {
    const wrapper = shallow(
      <Radio.Group shape="radius">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.hasClass('shape-radius')).toEqual(true);
  });

  // 椭圆角
  it('shape is round', () => {
    const wrapper = shallow(
      <Radio.Group shape="round">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.hasClass('shape-round')).toEqual(true);
  });

  // 块级样式
  it('block', () => {
    const wrapper = shallow(
      <Radio.Group block>
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2">选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.hasClass('block')).toEqual(true);
  });

  // 列表样式
  it('type is cell', () => {
    const wrapper = shallow(
      <Radio.Group type="cell">
        <Radio value="0">选项一</Radio>
        <Radio value="1">选项二</Radio>
        <Radio value="2" disabled>选项三</Radio>
      </Radio.Group>
    );
    expect(wrapper.props().children[0].props.type).toEqual('cell');
  });

  // 列表样式
  // it('type is cell and all options is disabled', () => {
  //   const wrapper = shallow(
  //     <Radio.Group type="cell" disabled>
  //       <Radio value="0">选项一</Radio>
  //       <Radio value="1">选项二</Radio>
  //       <Radio value="2">选项三</Radio>
  //     </Radio.Group>
  //   );
  //   expect(wrapper.props().children[0].props.type).toEqual('cell');
  // });
});
