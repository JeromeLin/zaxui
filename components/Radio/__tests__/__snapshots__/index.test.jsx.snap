// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly 1`] = `
<div
  class="za-radio-group"
>
  <a
    aria-disabled="false"
    class="za-button za-radio theme-primary size-xs"
    role="button"
  >
    <span>
      <input
        checked=""
        class="za-radio-input"
        type="radio"
        value="0"
      />
      选项一
    </span>
  </a>
  <a
    aria-disabled="false"
    class="za-button za-radio theme-primary size-xs bordered"
    role="button"
  >
    <span>
      <input
        class="za-radio-input"
        type="radio"
        value="1"
      />
      选项二
    </span>
  </a>
  <a
    aria-disabled="false"
    class="za-button za-radio theme-primary size-xs bordered"
    role="button"
  >
    <span>
      <input
        class="za-radio-input"
        type="radio"
        value="2"
      />
      选项三
    </span>
  </a>
</div>
`;
